api/v1/invitations/PKPInvitationController.php

```diff
@@ -0,0 +1,236 @@
+            ->name('invitation.finalize')
+            ->whereNumber('invitationId');
+        
+        Route::put('{invitationId}/key/{key}/refine', $this->refine(...))
+            ->name('invitation.refine')
+            ->whereNumber('invitationId');
+        
+        Route::put('{invitationId}/key/{key}/decline', $this->decline(...))
+            ->name('invitation.decline')
+            ->whereNumber('invitationId');
+    }
+
+    /**
+     * @copydoc \PKP\core\PKPBaseController::authorize()
+     */
+    public function authorize(PKPRequest $request, array &$args, array $roleAssignments): bool
+    {
+        $illuminateRequest = $args[0]; /** @var \Illuminate\Http\Request $illuminateRequest */
+        $actionName = static::getRouteActionName($illuminateRequest);
+
+        $invitation = null;
+
+        $invitationType = $this->getParameter(self::PARAM_TYPE);
+        $invitationId = (int) $this->getParameter(self::PARAM_ID);
+        $invitationKey = $this->getParameter(self::PARAM_KEY);
+
+        if (in_array($actionName, $this->requiresType)) {
+            if (!isset($invitationType)) {
+                throw new Exception("Invitation Type parameter with the name '" . self::PARAM_TYPE . "' need to be declared");
+            }
+
+            $invitation = app(PKPInvitation::class)->create($invitationType);
+        } elseif (in_array($actionName, $this->requiresOnlyId)) {
+            if (!isset($invitationId)) {
+                throw new Exception("Invitation Id parameter with the name '" . self::PARAM_ID . "' need to be declared");
+            }
+
+            $invitationModel = Invitation::find($invitationId);
+            if (!isset($invitationModel)) {
+                throw new Exception("Invitation not found");
+            }
+
+            $invitation = app(PKPInvitation::class)->create($invitationModel->type, $invitationModel);
+        } elseif (in_array($actionName, $this->requiresIdAndKey)) {
+            if (!isset($invitationId) || !isset($invitationKey)) {
+                throw new Exception("Invitation Id parameter with the name '" . self::PARAM_ID . "' and invitation Key parameter with the name '" . self::PARAM_KEY . "' need to be declared");
+            }
+
+            $invitation = Repo::invitation()->getByIdAndKey($invitationId, $invitationKey);
+        }
+
+        if (!isset($invitation)) {
+            throw new Exception("Invitation could not be created");
+        }
+
+        $this->invitation = $invitation;
+
+        if (!array_key_exists(IApiHandleable::class, class_implements($this->invitation))) {
+            throw new Exception('This inviatation does not support API handling');
+        }
+
+        $this->createInvitationHandler = $invitation->getCreateInvitationController();
+        $this->receiveInvitationHandler = $invitation->getReceiveInvitationController();
+
+        if (in_array($actionName, $this->requiresType) || in_array($actionName, $this->requiresOnlyId)) {
```

Comment:
You are right. I changed the code so that in the authorization stage, a `$this->selectedHandler` object is being created, and then this is the object that is being used through the controller (authorization and endpoints alike)